eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/AsyncPaymentTriggerer.scala

```diff
@@ -0,0 +1,122 @@
+
+/**
+ * This actor waits for an async payment receiver to become ready to receive a payment or for a block timeout to expire.
+ * If the receiver of the payment is a connected peer, spawn a PeerReadyNotifier actor.
+ * TODO: If the receiver is not a connected peer, wait for a `ReceiverReady` onion message containing the specified paymentHash.
+ */
+
+object AsyncPaymentTriggerer {
+  // @formatter:off
+  sealed trait Command
+  case class Start(switchboard: ActorRef[Switchboard.GetPeerInfo]) extends Command
+  case class Watch(replyTo: ActorRef[Result], remoteNodeId: PublicKey, paymentHash: ByteVector32, timeout: BlockHeight) extends Command
+  private case class WrappedPeerReadyResult(result: PeerReadyNotifier.Result) extends Command
+  private case class WrappedCurrentBlockHeight(currentBlockHeight: CurrentBlockHeight) extends Command
+
+  sealed trait Result
+  case object AsyncPaymentTriggered extends Result
+  case object AsyncPaymentTimeout extends Result
+  // @formatter:on
+
+  def apply(): Behavior[Command] = Behaviors.setup { context =>
+    new AsyncPaymentTriggerer(context).initializing()
+  }
+}
+
+private class AsyncPaymentTriggerer(context: ActorContext[Command]) {
+  import AsyncPaymentTriggerer._
+
+  case class Watcher(replyTo: ActorRef[Result], timeout: BlockHeight, paymentHash: ByteVector32) {
+    def expired(currentBlockHeight: BlockHeight): Boolean = timeout <= currentBlockHeight
+  }
+  case class AsyncPaymentTrigger(notifier: ActorRef[PeerReadyNotifier.Command], watchers: Set[Watcher]) {
+    def update(currentBlockHeight: BlockHeight): Option[AsyncPaymentTrigger] = {
+      // notify watchers that timeout occurred before offline peer reconnected
+      val expiredWatchers = watchers.filter(_.expired(currentBlockHeight))
+      expiredWatchers.foreach(e => e.replyTo ! AsyncPaymentTimeout)
+      // remove timed out watchers from set
+      val updatedWatchers: Set[Watcher] = watchers.removedAll(expiredWatchers)
+      if (updatedWatchers.isEmpty) {
+        // stop notifier for offline peer when all watchers time out
+        context.stop(notifier)
+        None
+      } else {
+        Some(AsyncPaymentTrigger(notifier, updatedWatchers))
+      }
+    }
+    def trigger(): Unit = watchers.foreach(e => e.replyTo ! AsyncPaymentTriggered)
+  }
+
+  private def initializing(): Behavior[Command] = {
+    Behaviors.receiveMessage[Command] {
+      case Start(switchboard) => watching(switchboard, Map())
+      case m => context.log.error(s"received unhandled message ${m.getClass.getSimpleName} before Start received.")
+        Behaviors.same
+    }
+  }
+
+  private def watching(switchboard: ActorRef[Switchboard.GetPeerInfo], triggers: Map[PublicKey, AsyncPaymentTrigger]): Behavior[Command] = {
+    val peerReadyResultAdapter = context.messageAdapter[PeerReadyNotifier.Result](WrappedPeerReadyResult)
+    context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[CurrentBlockHeight](WrappedCurrentBlockHeight))
```

Comment:
You are re-subscribing to the event stream every time you update the internal state and call `watching`, which isn't what we want. I extract that to be done in the initialization in https://github.com/ACINQ/eclair/commit/ac5118b0b5dc1c1f543bca076a43aba5f4ea9ef3